compiler/rustc_mir/src/transform/remove_unneeded_drops.rs

```diff
@@ -0,0 +1,57 @@
+//! This pass replaces a drop of a type that does not need dropping, with a goto
+
+use crate::transform::{MirPass, MirSource};
+use rustc_hir::def_id::LocalDefId;
+use rustc_middle::mir::visit::Visitor;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+
+pub struct RemoveUnneededDrops {
+    def_id: LocalDefId,
+}
+
+impl RemoveUnneededDrops {
+    pub fn new(def_id: LocalDefId) -> Self {
+        Self { def_id }
+    }
+}
+
+impl<'tcx> MirPass<'tcx> for RemoveUnneededDrops {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, source: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        trace!("Running RemoveUnneededDrops on {:?}", source);
+        let mut opt_finder = RemoveUnneededDropsOptimizationFinder {
+            tcx,
+            body,
+            optimizations: vec![],
+            def_id: self.def_id,
+        };
+        opt_finder.visit_body(body);
+        for (loc, target) in opt_finder.optimizations {
+            let terminator = body.basic_blocks_mut()[loc.block].terminator_mut();
+            debug!("SUCCESS: replacing `drop` with goto({:?})", target);
+            terminator.kind = TerminatorKind::Goto { target };
+        }
```

Comment:
We have a goto chain collapser. I think we should run it if `!opt_finder.optimizations.is_empty()` so that follow up optimizations get the best code to go forward with